\documentclass{resume} % Use the custom resume.cls style
\usepackage[left=0.75in,top=0.6in,right=0.75in,bottom=0.6in]{geometry} % Document margins
\usepackage{xcolor}
\usepackage{tabularx}
\name{ test2 } % Your name
\address{Generated on \today} % Your address
\begin{document}

\begin{rSection}{ acenocoumarol }
\item[]
\begin{rSubsection}{ CYP2C9 }{ cytochrome P450, family 2, subfamily C, polypeptide 9 }{}{}
\item[]
\item[] ------------------------------------------------------ Dosing Guideline --------------------------------------------------------\newline
\item[]
\item[] \textbf{ *18/*1 } | \textbf{ *3/*1 }
\item The Royal Dutch Pharmacists Association - Pharmacogenetics Working Group has evaluated therapeutic dose recommendations for acenocoumarol based on CYP2C9 genotype PMID:21412232.
 \newline
\vspace{1pt}\newline
		\scriptsize
		\begin{center}
		\begin{tabularx}{0.9\textwidth}{ bsss }
		\textbf{ Genotype }&\textbf{ Therapeutic Dose Recommendation }&\textbf{ Level of Evidence }&\textbf{ Clinical Relevance }\\
		\vspace{1pt}\\
		\hline \\
		\vspace{1pt}\\
		         CYP2C9 *1/*2 & Check INR more frequently after initiating or discontinuing NSAIDs & Published controlled studies of good quality* relating to phenotyped and/or genotyped patients or healthy volunteers, and having relevant pharmacokinetic or clinical endpoints. & Minor clinical effect (S): QTc prolongation (:450 ms females, :470 ms males),  INR increase :4.5. Kinetic effect (S) \\
		\vspace{1pt}\\
		\hline \\
		\vspace{1pt}\\
		         CYP2C9 *2/*2 & Check INR more frequently after initiating or discontinuing NSAIDs & Published controlled studies of good quality* relating to phenotyped and/or genotyped patients or healthy volunteers, and having relevant pharmacokinetic or clinical endpoints. & Minor clinical effect (S): QTc prolongation (:450 ms females, :470 ms males),  INR increase :4.5. Kinetic effect (S) 
\\
		\vspace{1pt}\\
		\hline \\
		\vspace{1pt}\\
		         CYP2C9 *1/*3 & Check INR more frequently after initiating or discontinuing NSAIDs & Published controlled studies of good quality* relating to phenotyped and/or genotyped patients or healthy volunteers, and having relevant pharmacokinetic or clinical endpoints. & Clinical effect (S): short-lived discomfort (: 48 hr) without permanent injury: e.g. reduced decrease in resting heart rate,  reduction in exercise tachycardia,  decreased pain relief from oxycodone,  ADE resulting from increased bioavailability of atomoxetine (decreased appetite, insomnia, sleep disturbance etc),  neutropenia ,  1.5x109/l,  leucopenia ,  3.0x109/l,  thrombocytopenia  ,  75x109/l,  moderate diarrhea not affecting daily activities,  reduced glucose increase following oral glucose tolerance test. 
\\
		\vspace{1pt}\\
		\hline \\
		\vspace{1pt}\\
		         CYP2C9 *2/*3 & Check INR more frequently after initiating or discontinuing NSAIDs & Published controlled studies of good quality* relating to phenotyped and/or genotyped patients or healthy volunteers, and having relevant pharmacokinetic or clinical endpoints. & Minor clinical effect (S): QTc prolongation (:450 ms females, :470 ms males),  INR increase :4.5. Kinetic effect (S) 
\\
		\vspace{1pt}\\
		\hline \\
		\vspace{1pt}\\
		         CYP2C9 *3/*3 & Check INR more frequently during dose titration and after initiating or discontinuing NSAIDs & Published controlled studies of good quality* relating to phenotyped and/or genotyped patients or healthy volunteers, and having relevant pharmacokinetic or clinical endpoints. & Minor clinical effect (S): QTc prolongation (:450 ms females, :470 ms males),  INR increase :4.5. Kinetic effect (S) \\
\end{tabularx}
\end{center}
\normalsize
\vspace{10pt}
		        
\item[] ---------------------------------------------------- Clinical Annotations -----------------------------------------------------\newline
\item \textbf{\colorbox{cyan} {Class 2A}} \textbf{ rs1057910 } \textit{ AC }
\item[] Patients with the AC genotype may require decreased dose of acenocoumarol or closer INR monitoring as compared to patients with the AA genotype. Other genetic and clinical factors may also influence acenocoumarol dose.
\end{rSubsection}
\end{rSection}\begin{rSubsection}{ VKORC1 }{ vitamin K epoxide reductase complex, subunit 1 }{}{}
\item[]
\item[] ------------------------------------------------------ Dosing Guideline --------------------------------------------------------\newline
\item[]
\item[] \textbf{ H2/H1 } | \textbf{ H2/H1 }
\item The Royal Dutch Pharmacists Association - Pharmacogenetics Working Group has evaluated therapeutic dose recommendations for acenocoumarol based on VKORC1 genotype PMID:21412232.  They found that VKORC1 genotype contributes to dose variability.  However, they make no dosing recommendations at this time , because of strict international normalized ratio monitoring by the Dutch Thrombosis Service., 
 \newline
\vspace{1pt}\newline
		\scriptsize
		\begin{center}
		\begin{tabularx}{0.9\textwidth}{ bssss }
		\textbf{ Genotype }&\textbf{ Therapeutic Dose Recommendation }&\textbf{ Level of Evidence }&\textbf{ Clinical Relevance }&\textbf{
}\\
		\vspace{1pt}\\
		\hline \\
		\vspace{1pt}\\
		         VKORC1 variant:rs9934438 AG & None & Published controlled studies of good quality* relating to phenotyped and/or genotyped patients or healthy volunteers, and having relevant pharmacokinetic or clinical endpoints. & Minor clinical effect (S): QTc prolongation (:450 ms , :470 ms ),  INR increase : 4.5,  Kinetic effect (S). & 
\\
		\vspace{1pt}\\
		\hline \\
		\vspace{1pt}\\
		         VKORC1 variant:rs9934438 AA & Check INR more frequently. & Published controlled studies of good quality* relating to phenotyped and/or genotyped patients or healthy volunteers, and having relevant pharmacokinetic or clinical endpoints. & Minor clinical effect (S): QTc prolongation (:450 ms , :470 ms ),  INR increase : 4.5,  Kinetic effect (S). &
\\
		\end{tabularx}
		\end{center}
		\normalsize
		\vspace{10pt}
		        
\item[] ---------------------------------------------------- Clinical Annotations -----------------------------------------------------\newline
\item \textbf{\colorbox{cyan} {Class 2A}} \textbf{ rs9934438 } \textit{ GA }
\item[] Patients with the AG genotype may have decreased dose of acenocoumarol or phenprocoumon as compared to patients with genotype GG. Other genetic and clinical factors may also influence the dose of acenocoumarol or phenprocoumon.
\end{rSubsection}

\end{rSection}\begin{rSection}{ antipsychotics }
\item[]
\begin{rSection}{ atazanavir }
\item[]
\begin{rSubsection}{ UGT1A1 }{ UDP glucuronosyltransferase 1 family, polypeptide A1 }{}{}
\item[]
\item[] ------------------------------------------------------ Dosing Guideline --------------------------------------------------------\newline
\item[]
\item[] \textbf{ *60/*1 } | \textbf{ *60/*1 }
\item April 2016
 \newline
\item Advance online publication September 2015
 \newline
\item Adults
 \newline
\item At the time of this writing there are no pediatric data regarding associations between UGT1A1 genotypes and likelihood of bilirubin-related discontinuation of atazanavir. However, UGT1A1 genotypes are expected to affect atazanavir-related hyperbilirubinemia similarly in adults and children. Therefore, recommendations for adults may be directly adapted to pediatric patients.
 \newline
\item Clinical Pharmacogenetics Implementation Consortium Guidelines for Atazanavir and UGT1A1 Prescribing(https://github.com/PharmGKB/cpic-guidelines/raw/master/atazanavir/2015/26417955.pdf).
 \newline
\item 2015 Supplement(https://github.com/PharmGKB/cpic-guidelines/raw/master/atazanavir/2015/26417955-supplement.pdf)
 \newline
\item UGT1A1 Allele Frequencies(https://github.com/PharmGKB/cpic-guidelines/raw/master/atazanavir/2015/26417955-UGT1A1allelefrequency.xlsx)
 \newline
\item Table 1: Recommended therapeutic use of atazanavir based on UGT1A1 genotype
 \newline
\item Adapted from Tables 1 and 2 of the 2015 guideline manuscript.
 \newline
\vspace{1pt}\newline
		\scriptsize
		\begin{center}
		\begin{tabularx}{0.9\textwidth}{ bssssss }
		\textbf{ Likely phenotype }&\textbf{ Genotypes }&\textbf{ Examples of diplotypes }&\textbf{ Implications for phenotypic measures   }&\textbf{ Recommendations for atazanavir therapy }&\textbf{ Classification of recommendation for atazanavir therapy }&\textbf{
}\\
		\vspace{1pt}\\
		\hline \\
		\vspace{1pt}\\
		        Extensive Metabolizer & An individual carrying 2 reference b function and/or increased function alleles,  or individuals of genotype CC at variant:rs887829 & *1/*1,  *1/*36,  *36/*36,  variant:rs887829 CC& Reference c UGT1A1 activity,  very low likelihood of bilirubin-related discontinuation of atazanavir.  &  There is no need to avoid prescribing of atazanavir based on UGT1A1 genetic test result.  &  Strong &
\\
		\vspace{1pt}\\
		\hline \\
		\vspace{1pt}\\
		        Intermediate Metabolizer  & An individual carrying one reference b function (*1) c or increased function allele (*36) plus one decreased function allele (*6, *28, *37). Alternatively identified by heterozygosity for variant:rs887829 C/T. & *1/*28,  *1/*37,  *36/*28,  *36/*37,  variant:rs887829 C/T, *1/*6 & Somewhat decreased UGT1A1 activity,  low likelihood of bilirubin-related discontinuation of atazanavir. & There is no need to avoid prescribing of atazanavir based on UGT1A1 genetic test result. Inform the patient that some patients stop atazanavir because of jaundice (yellow eyes and skin), but that this patient’s genotype makes this unlikely &  Strong &
\\
		\vspace{1pt}\\
		\hline \\
		\vspace{1pt}\\
		        Poor Metabolizer  & An individual carrying two decreased function alleles (*6, *28, *37). Alternatively identified by homozygosity for variant:rs887829 T/T (*80/*80) & *28/*28,  *28/*37,  *37/*37,  variant:rs887829 T/T (*80/*80), *6/*6 a & Markedly decreased UGT1A1 activity,  high likelihood of bilirubin-related discontinuation of atazanavir. & Consider an alternative agent particularly where jaundice would be of concern to the patient. &  Strong \\
		\end{tabularx}
		\end{center}
		\normalsize
		\vspace{10pt}
		        
\begin{rSection}{ capecitabine }
\item[]
\begin{rSubsection}{ DPYD }{ dihydropyrimidine dehydrogenase }{}{}
\item[]
\item[] ------------------------------------------------------ Dosing Guideline --------------------------------------------------------\newline
\item[]
\item[] \textbf{ *1/*1 } | \textbf{ *1/*1 }
\item May 2014 Update on PharmGKB
 \newline
\item December 2013 Publication
 \newline
\item Accepted article preview online August 2013,  Advance online publication October 2013.
 \newline
\item at the time of this writing, there are no data available on the possible role of DPYD*2A, *13, or variant:rs67376798 in 5-fluorouracil toxicities in pediatric patient populations,  however, there is no reason to suspect that DPYD variant alleles would affect 5-fluorouracil metabolism differently in children compared to adults.
 \newline
\item The strength of the dosing recommendations is based on the fact that some variants (DPYD*2A, *13, and variant:rs67376798) clearly affect DPD activity, and DPD activity is clearly related to 5-fluorouracil clearance, and 5-fluorouracil exposure is associated with its toxic effects. Therefore, reduction of fluoropyrimidine dosage in patients with these variants may prevent severe and possibly life-threatening toxicities. However, available evidence does not clearly indicate a degree of dose reduction needed to prevent fluoropyrimidine related toxicities...Based on literature review (see full manuscript), our recommendation is to start with at least a 50\% reduction of the starting dose followed by an increase in dose in patients experiencing no or clinically tolerable toxicity to maintain efficacy, a decrease in dose in patients who do not tolerate the starting dose to minimize toxicities or pharmacokinetic guided dose adjustments (if available). Patients who are homozygous for DPYD*2A, *13, or variant:rs67376798 may demonstrate complete DPD deficiency and the use of 5-fluorouracil or capecitabine is not recommended in these patients.,  
 \newline
\item Clinical Pharmacogenetics Implementation Consortium Guidelines for Dihydropyrimidine Dehydrogenase Genotype and Fluoropyrimidine Dosing(https://github.com/PharmGKB/cpic-guidelines/raw/master/fluoropyrimidines/2013/23988873.pdf)
 \newline
\item 2013 supplement(https://github.com/PharmGKB/cpic-guidelines/raw/master/fluoropyrimidines/2013/23988873-supplement.pdf)
 \newline
\item Table 1: Recommended dosing of fluoropyrimidines by genotype/phenotype.
 \newline
\item Adapted from Tables 1 and 2 of the 2013 guideline manuscript.
 \newline
\vspace{1pt}\newline
		\scriptsize
		\begin{center}
		\begin{tabularx}{0.9\textwidth}{ bsssss }
		\textbf{ Phenotype (genotype) }&\textbf{ Examples of diplotypes }&\textbf{ Implications for phenotypic measures }&\textbf{ Dosing recommendations }&\textbf{ Classification of recommendations a }&\textbf{
}\\
		\vspace{1pt}\\
		\hline \\
		\vspace{1pt}\\
		         Homozygous wild-type or normal, high DPD activity (two or more functional *1 alleles) & *1/*1 & Normal DPD activity and , normal,  risk for fluoropyrimidine toxicity & Use label-recommended dosage and administration & Moderate &
\\
		\vspace{1pt}\\
		\hline \\
		\vspace{1pt}\\
		         Heterozygous or intermediate activity (~3-5\% of patients), may have partial DPD deficiency, at risk for toxicity with drug exposure (one functional allele *1, plus one nonfunctional allele - *2A, *13 or rs67376798A c) & *1/*2A,  *1/*13,  *1/ rs67376798A c) & Decreased DPD activity (leukocyte DPD activity at 30\% to 70\% that of the normal population) and increased risk for severe or even fatal drug toxicity when treated with fluoropyrimidine drugs & Start with at least a 50\% reduction in starting dose followed by titration of dose based on toxicity b or pharmacokinetic test (if available) & Moderate &
\\
		\vspace{1pt}\\
		\hline \\
		\vspace{1pt}\\
		         Homozygous variant, DPD deficiency (~0.2\% of patients), at risk for toxicity with drug exposure (2 nonfunctional alleles - *2A, *13 or rs67376798A c) & *2A/*2A,  *13/*13,  rs67376798A c / rs67376798A c & Complete DPD deficiency and increased risk for severe or even fatal drug toxicity when treated with fluoropyrimidine drugs & Select alternate drug & Strong &
\\
		\end{tabularx}
		\end{center}
		\normalsize
		\vspace{10pt}
		        
\item[] ---------------------------------------------------- Clinical Annotations -----------------------------------------------------\newline
\item \textbf{\colorbox{red} {Class 1A}} \textbf{ rs55886062 } \textit{ AA }
\item[] Patients with the AA genotype (DPYD *1/*1) and cancer who are treated with fluoropyrimidine-based chemotherapy may have a decreased, but not absent, risk for drug toxicity as compared to patients with the AC or CC genotype (DPYD *1/*13 or *13/*13). Fluoropyrimidines are often used in combination chemotherapy such as FOLFOX (fluorouracil, leucovorin and oxaliplatin), FOLFIRI (fluorouracil, leucovorin and irinotecan) or FEC (fluorouracil, epirubicin and cyclophosphamide) or with other drugs such as bevacizumab, cetuximab, raltitrexed. The combination and delivery of the drug may influence risk for toxicity. Other genetic and clinical factors may also influence response to fluoropyrimidine-based chemotherapy.\item \textbf{\colorbox{red} {Class 1A}} \textbf{ rs3918290 } \textit{ CC }
\item[] Patients with the CC genotype (DPYD *1/*1) and cancer who are treated with fluoropyrimidine-based chemotherapy may have 1) increased clearance of fluoropyrimidine drugs and 2) decreased, but not non-existent, risk for drug toxicity as compared to patients with the CT or TT genotype (DPYD *1/*2A or *2A/*2A). Fluoropyrimidines are often used in combination chemotherapy such as FOLFOX (fluorouracil, leucovorin and oxaliplatin), FOLFIRI (fluorouracil,  leucovorin and irinotecan) or FEC (fluorouracil, epirubicin and cyclophosphamide) or with other drugs such as bevacizumab, cetuximab, raltitrexed. The combination and delivery of the drug may influence risk for toxicity. Other genetic and clinical factors may also influence response to fluoropyrimidine based chemotherapy.\item \textbf{\colorbox{red} {Class 1A}} \textbf{ rs67376798 } \textit{ TT }
\item[] Patients with the TT genotype and cancer who are treated with fluoropyrimidine-based chemotherapy may have 1) increased clearance of the drug and 2) decreased, but not absent, risk and reduced severity of drug toxicity as compared to patients with the AT genotype. Fluoropyrimidines are often used in combination chemotherapy such as FOLFOX (fluorouracil, leucovorin and oxaliplatin), FOLFIRI (fluorouracil, leucovorin and irinotecan) or FEC (fluorouracil, epirubicin and cyclophosphamide) or with other drugs such as bevacizumab, cetuximab, raltitrexed. The combination and delivery of the drug may influence risk for toxicity. Other genetic and clinical factors may also influence response to fluoropyrimidine-based chemotherapy.
\end{rSubsection}

\begin{rSection}{ fluorouracil }
\item[]
\begin{rSubsection}{ DPYD }{ dihydropyrimidine dehydrogenase }{}{}
\item[]
\item[] ------------------------------------------------------ Dosing Guideline --------------------------------------------------------\newline
\item[]
\item[] \textbf{ *1/*1 } | \textbf{ *1/*1 }
\item May 2014 Update on PharmGKB
 \newline
\item December 2013 Publication
 \newline
\item Accepted article preview online August 2013,  Advance online publication October 2013.
 \newline
\item at the time of this writing, there are no data available on the possible role of DPYD*2A, *13, or variant:rs67376798 in 5-fluorouracil toxicities in pediatric patient populations,  however, there is no reason to suspect that DPYD variant alleles would affect 5-fluorouracil metabolism differently in children compared to adults.
 \newline
\item The strength of the dosing recommendations is based on the fact that some variants (DPYD*2A, *13, and variant:rs67376798) clearly affect DPD activity, and DPD activity is clearly related to 5-fluorouracil clearance, and 5-fluorouracil exposure is associated with its toxic effects. Therefore, reduction of fluoropyrimidine dosage in patients with these variants may prevent severe and possibly life-threatening toxicities. However, available evidence does not clearly indicate a degree of dose reduction needed to prevent fluoropyrimidine related toxicities...Based on literature review (see full manuscript), our recommendation is to start with at least a 50\% reduction of the starting dose followed by an increase in dose in patients experiencing no or clinically tolerable toxicity to maintain efficacy, a decrease in dose in patients who do not tolerate the starting dose to minimize toxicities or pharmacokinetic guided dose adjustments (if available). Patients who are homozygous for DPYD*2A, *13, or variant:rs67376798 may demonstrate complete DPD deficiency and the use of 5-fluorouracil or capecitabine is not recommended in these patients.,  
 \newline
\item Clinical Pharmacogenetics Implementation Consortium Guidelines for Dihydropyrimidine Dehydrogenase Genotype and Fluoropyrimidine Dosing(https://github.com/PharmGKB/cpic-guidelines/raw/master/fluoropyrimidines/2013/23988873.pdf)
 \newline
\item 2013 supplement(https://github.com/PharmGKB/cpic-guidelines/raw/master/fluoropyrimidines/2013/23988873-supplement.pdf)
 \newline
\item Table 1: Recommended dosing of fluoropyrimidines by genotype/phenotype.
 \newline
\item Adapted from Tables 1 and 2 of the 2013 guideline manuscript.
 \newline
\vspace{1pt}\newline
		\scriptsize
		\begin{center}
		\begin{tabularx}{0.9\textwidth}{ bsssss }
		\textbf{ Phenotype (genotype) }&\textbf{ Examples of diplotypes }&\textbf{ Implications for phenotypic measures }&\textbf{ Dosing recommendations }&\textbf{ Classification of recommendationsa }&\textbf{
}\\
		\vspace{1pt}\\
		\hline \\
		\vspace{1pt}\\
		         Homozygous wild-type or normal, high DPD activity (two or more functional *1 alleles) & *1/*1 & Normal DPD activity and , normal,  risk for fluoropyrimidine toxicity & Use label-recommended dosage and administration & Moderate &
\\
		\vspace{1pt}\\
		\hline \\
		\vspace{1pt}\\
		         Heterozygous or intermediate activity (~3-5\% of patients), may have partial DPD deficiency, at risk for toxicity with drug exposure (one functional allele *1, plus one nonfunctional allele - *2A, *13 or rs67376798Ac) & *1/*2A,  *1/*13,  *1/ rs67376798Ac) & Decreased DPD activity (leukocyte DPD activity at 30\% to 70\% that of the normal population) and increased risk for severe or even fatal drug toxicity when treated with fluoropyrimidine drugs & Start with at least a 50\% reduction in starting dose followed by titration of dose based on toxicity b or pharmacokinetic test (if available) & Moderate &
\\
		\vspace{1pt}\\
		\hline \\
		\vspace{1pt}\\
		         Homozygous variant, DPD deficiency (~0.2\% of patients), at risk for toxicity with drug exposure (2 nonfunctional alleles - *2A, *13 or rs67376798A c) & *2A/*2A,  *13/*13,  rs67376798Ac / rs67376798Ac & Complete DPD deficiency and increased risk for severe or even fatal drug toxicity when treated with fluoropyrimidine drugs & Select alternate drug & Strong &
\\
		\end{tabularx}
		\end{center}
		\normalsize
		\vspace{10pt}
		        
\item[] ---------------------------------------------------- Clinical Annotations -----------------------------------------------------\newline
\item \textbf{\colorbox{red} {Class 1A}} \textbf{ rs55886062 } \textit{ AA }
\item[] Patients with the AA genotype (DPYD *1/*1) and cancer who are treated with fluoropyrimidine-based chemotherapy may have a decreased, but not absent, risk for drug toxicity as compared to patients with the AC or CC genotype (DPYD *1/*13 or *13/*13). Fluoropyrimidines are often used in combination chemotherapy such as FOLFOX (fluorouracil, leucovorin and oxaliplatin), FOLFIRI (fluorouracil, leucovorin and irinotecan) or FEC (fluorouracil, epirubicin and cyclophosphamide) or with other drugs such as bevacizumab, cetuximab, raltitrexed. The combination and delivery of the drug may influence risk for toxicity. Other genetic and clinical factors may also influence response to fluoropyrimidine-based chemotherapy.\item \textbf{\colorbox{red} {Class 1A}} \textbf{ rs3918290 } \textit{ CC }
\item[] Patients with the CC genotype (DPYD *1/*1) and cancer who are treated with fluoropyrimidine-based chemotherapy may have 1) increased clearance of fluoropyrimidine drugs and 2) decreased, but not non-existent, risk for drug toxicity as compared to patients with the CT or TT genotype (DPYD *1/*2A or *2A/*2A). Fluoropyrimidines are often used in combination chemotherapy such as FOLFOX (fluorouracil, leucovorin and oxaliplatin), FOLFIRI (fluorouracil,  leucovorin and irinotecan) or FEC (fluorouracil, epirubicin and cyclophosphamide) or with other drugs such as bevacizumab, cetuximab, raltitrexed. The combination and delivery of the drug may influence risk for toxicity. Other genetic and clinical factors may also influence response to fluoropyrimidine based chemotherapy.\item \textbf{\colorbox{red} {Class 1A}} \textbf{ rs67376798 } \textit{ TT }
\item[] Patients with the TT genotype and cancer who are treated with fluoropyrimidine-based chemotherapy may have 1) increased clearance of the drug and 2) decreased, but not absent, risk and reduced severity of drug toxicity as compared to patients with the AT genotype. Fluoropyrimidines are often used in combination chemotherapy such as FOLFOX (fluorouracil, leucovorin and oxaliplatin), FOLFIRI (fluorouracil, leucovorin and irinotecan) or FEC (fluorouracil, epirubicin and cyclophosphamide) or with other drugs such as bevacizumab, cetuximab, raltitrexed. The combination and delivery of the drug may influence risk for toxicity. Other genetic and clinical factors may also influence response to fluoropyrimidine-based chemotherapy.
\end{rSubsection}
\end{rSection}

\begin{rSubsection}{ UGT1A1 }{ UDP glucuronosyltransferase 1 family, polypeptide A1 }{}{}
\item[]
\item[] ------------------------------------------------------ Dosing Guideline --------------------------------------------------------\newline
\item[]
\item[] \textbf{ *60/*1 } | \textbf{ *60/*1 }
\item The French joint working group comprising the National Pharmacogenetics Network (RNPGx) and the Group of Clinical Onco-pharmacology (GPCO-Unicancer) has published UGT1A1-based drug dosing guidelines for irinotecan in Fundamental &amp,  Clinical Pharmacology. Excerpts from , UGT1A1 genotype and irinotecan therapy: general review and implementation in routine practice,  PMID:25817555 follow:
 \newline
\item ...we recommend pretreatment UGT1A1 genotyping of the TATA box (*28, *36, *37) for all patients scheduled to receive an irinotecan dose , =180 mg/m2. The rare allele *36 (proficient) can be interpreted as an *1 allele and allele *37 (deficient) as an allele *28.
 \newline
\item For low irinotecan doses (:180 mg/m2/week) UGT1A1 genotyping is not indicated as hematological and gastrointestinal toxicities are quite similar regardless of the genotype.
 \newline
\item For initially scheduled doses between 180 and 230 mg/m2 every 2-3 weeks, *28/*28 patients are at increased risk of developing hematological and/or digestive toxicity as compared to other genotypes...a 25-30\% dose reduction at the first cycle is recommended, particularly in cases of associated risk factors (performance status , 3).
 \newline
\item For initially scheduled doses , =240 mg/m2 every 2-3 weeks, *28/*28 patients are at a much higher risk of hematological toxicity (neutropenia) as compared to other genotypes. We thus recommend contraindicating such an intensified dose in *28/*28 patients. The administration of an intensified dose (240 mg/m2) is only possible in *1/*1 patients, as well as in *1/*28 patients, in the absence of additional risk factors and under strict medical surveillance.
 \newline
\item This...analysis is limited by the fact that other UGT1A1 deficient variants are relevant in non-Caucasian populations, particularly the *6 and *27 alleles in Asian populations.
 \newline
\item The joint working group also provided a decision tree to guide irinotecan dosing based on UGT1A1 genotype:
 \newline
\item !Decision tree for irinotecan dosing(Decisiontreeforirinotecandosing.png)
 \newline
\item Reprinted with permission from Etienne-Grimaldi et al. UGT1A1 genotype and irinotecan therapy: general review and implementation in routine practice. Fundamental &amp,  Clinical Pharmacology (2015) \newline
\vspace{1pt}\newline


\begin{rSection}{ peginterferon alfa-2a }
\item[]
\begin{rSubsection}{ IFNL3 }{ interferon, lambda 3 }{}{}
\item[]
\item[] ------------------------------------------------------ Dosing Guideline --------------------------------------------------------\newline
\item[]
\item[] \textbf{ rs12979860T/rs12979860C } | \textbf{ rs12979860T/rs12979860C }
\item February 2014
 \newline
\item Accepted article preview online October 2013,  Advance online publication November 2013.
 \newline
\item adults
 \newline
\item The role of IFNL3 genotyping depends on treatment selection. IFNL3 genotype is only one factor that could influence response rates to PEG-IFN alpha and RBV therapy in HCV genotype 1 infection and should be interpreted in the context of other clinical and genetic factors. , 
 \newline
\item For patients treated with PEG-IFN alpha and RBV alone, IFNL3 genotype is the strongest pretreatment predictor of HCV treatment response. In the intention to treat analysis of the original discovery cohort with variant:rs12979860, Caucasian CC genotype patients were more likely than CT or TT patients to be undetectable by week 4 (28\% versus 5\% and 5\%, P:0.0001) and to achieve SVR (69\% versus 33\% and 27\%, P:0.0001). Similar patterns were observed in Hispanic and African American patients in this cohort. HCV treatment is associated with significant side effects, and the likelihood of response treatment influences shared decision making between clinicians and patients about initiating treatment.,  
 \newline
\item For treatment naïve patients with genotype 1 infection who are treated with protease inhibitor combinations, all IFNL3 genotypes have improved response rates compared to PEG-IFN alpha and RBV only. However, patients with the favorable IFNL3 genotype still have higher response rates with the protease inhibitor combination in treatment naïve patients, and these response rates may guide patients and clinicians in their treatment decisions. In the boceprevir phase 3 naïve study of combination with PEG-IFN alpha and RBV, SVR rates for variant:rs12979860 CC patients receiving boceprevir ranged from 80-82\% compared to 65-71\% for CT patients and 59-65\% for TT patients. Moreover, multivariate regression analysis revealed that variant:rs12979860 CC was a predictor of SVR compared to CT (odds ratio (OR) 2.6, 95\% CI 1.3-5.1) and TT genotypes (OR 2.1, 95\% CI 1.2-3.7)., 
 \newline
\item Clinical Pharmacogenetics Implementation Consortium (CPIC) guidelines for IFNL3 (IL28B) genotype and PEG interferon-alpha-based regimens(https://github.com/PharmGKB/cpic-guidelines/raw/master/pegintron/2013/24096968.pdf) 
 \newline
\item 2013 supplement(https://github.com/PharmGKB/cpic-guidelines/raw/master/pegintron/2013/24096968-supplement.pdf)
 \newline
\item Table 1: Recommended therapeutic use of  PEG-interferon-alpha containing regimens based on IFNL3 genotype
 \newline
\item Adapted from Table 1 and 2 of the 2013 guideline manuscript
 \newline
\item PEG-IFN alpha: pegylated-interferon alpha 2a or 2b,  RBV: ribavirin \newline
\vspace{1pt}\newline
		\scriptsize
		\begin{center}
		\begin{tabularx}{0.9\textwidth}{ bsssss }
		\textbf{ Genotype at variant:rs12979860 }&\textbf{ Phenotype }&\textbf{ Implications for PEG-IFN alpha and RBV a }&\textbf{ Implications for protease inhibitor combinations with PEG-IFN alpha and RBV therapy }&\textbf{ Classification of recommendations b }&\textbf{
}\\
		\vspace{1pt}\\
		\hline \\
		\vspace{1pt}\\
		         CC & Favorable response genotype & Approximately 70\% chance for SVR c after 48 weeks of treatment. Consider implications before initiating PEG-IFN alpha and RBV containing regimens. & Approximately 90\% chance for SVR after 24-48 weeks of treatment. Approximately 80-90\% of patients are eligible for shortened therapy (24-28 weeks vs. 48 weeks)d. Weighs in favor of using PEG-IFN alpha and RBV containing regimens.& Strong &
\\
		\vspace{1pt}\\
		\hline \\
		\vspace{1pt}\\
		         CT or TT & Unfavorable response genotype & Approximately 30\% chance for SVR c after 48 weeks of treatment. Consider implications before initiating PEG-IFN alpha and RBV containing regimens. & Approximately 60\% chance for SVR after 24-48 weeks of treatment. Approximately 50\% of patients are eligible for shortened therapy (24-28 weeks)d. Consider implications before initiating PEG-IFN and RBV containing regimens.& Strong &
\\
		\end{tabularx}
		\end{center}
		\normalsize
		\vspace{10pt}
		        

\end{rSubsection}
\end{rSection}\begin{rSection}{ peginterferon alfa-2b }
\item[]
\begin{rSubsection}{ IFNL3 }{ interferon, lambda 3 }{}{}
\item[]
\item[] ------------------------------------------------------ Dosing Guideline --------------------------------------------------------\newline
\item[]
\item[] \textbf{ rs12979860T/rs12979860C } | \textbf{ rs12979860T/rs12979860C }
\item February 2014
 \newline
\item Accepted article preview online October 2013,  Advance online publication November 2013.
 \newline
\item adults
 \newline
\item The role of IFNL3 genotyping depends on treatment selection. IFNL3 genotype is only one factor that could influence response rates to PEG-IFN alpha and RBV therapy in HCV genotype 1 infection and should be interpreted in the context of other clinical and genetic factors. , 
 \newline
\item For patients treated with PEG-IFN alpha and RBV alone, IFNL3 genotype is the strongest pretreatment predictor of HCV treatment response. In the intention to treat analysis of the original discovery cohort with variant:rs12979860, Caucasian CC genotype patients were more likely than CT or TT patients to be undetectable by week 4 (28\% versus 5\% and 5\%, P:0.0001) and to achieve SVR (69\% versus 33\% and 27\%, P:0.0001). Similar patterns were observed in Hispanic and African American patients in this cohort. HCV treatment is associated with significant side effects, and the likelihood of response treatment influences shared decision making between clinicians and patients about initiating treatment.,  
 \newline
\item For treatment naïve patients with genotype 1 infection who are treated with protease inhibitor combinations, all IFNL3 genotypes have improved response rates compared to PEG-IFN alpha and RBV only. However, patients with the favorable IFNL3 genotype still have higher response rates with the protease inhibitor combination in treatment naïve patients, and these response rates may guide patients and clinicians in their treatment decisions. In the boceprevir phase 3 naïve study of combination with PEG-IFN alpha and RBV, SVR rates for variant:rs12979860 CC patients receiving boceprevir ranged from 80-82\% compared to 65-71\% for CT patients and 59-65\% for TT patients. Moreover, multivariate regression analysis revealed that variant:rs12979860 CC was a predictor of SVR compared to CT (odds ratio (OR) 2.6, 95\% CI 1.3-5.1) and TT genotypes (OR 2.1, 95\% CI 1.2-3.7)., 
 \newline
\item Clinical Pharmacogenetics Implementation Consortium (CPIC) guidelines for IFNL3 (IL28B) genotype and PEG interferon-alpha-based regimens(https://github.com/PharmGKB/cpic-guidelines/raw/master/pegintron/2013/24096968.pdf) 
 \newline
\item 2013 supplement(https://github.com/PharmGKB/cpic-guidelines/raw/master/pegintron/2013/24096968-supplement.pdf)
 \newline
\item Table 1: Recommended therapeutic use of  PEG-interferon-alpha containing regimens based on IFNL3 genotype
 \newline
\item Adapted from Table 1 and 2 of the 2013 guideline manuscript
 \newline
\item PEG-IFN alpha: pegylated-interferon alpha 2a or 2b,  RBV: ribavirin \newline
\vspace{1pt}\newline
		\scriptsize
		\begin{center}
		\begin{tabularx}{0.9\textwidth}{ bsssss }
		\textbf{ Genotype at variant:rs12979860 }&\textbf{ Phenotype }&\textbf{ Implications for PEG-IFN alpha and RBV a }&\textbf{ Implications for protease inhibitor combinations with PEG-IFN alpha and RBV therapy }&\textbf{ Classification of recommendations b }&\textbf{
}\\
		\vspace{1pt}\\
		\hline \\
		\vspace{1pt}\\
		         CC & Favorable response genotype & Approximately 70\% chance for SVR c after 48 weeks of treatment. Consider implications before initiating PEG-IFN alpha and RBV containing regimens. & Approximately 90\% chance for SVR after 24-48 weeks of treatment. Approximately 80-90\% of patients are eligible for shortened therapy (24-28 weeks vs. 48 weeks)d. Weighs in favor of using PEG-IFN alpha and RBV containing regimens.& Strong &
\\
		\vspace{1pt}\\
		\hline \\
		\vspace{1pt}\\
		         CT or TT & Unfavorable response genotype & Approximately 30\% chance for SVR c after 48 weeks of treatment. Consider implications before initiating PEG-IFN alpha and RBV containing regimens. & Approximately 60\% chance for SVR after 24-48 weeks of treatment. Approximately 50\% of patients are eligible for shortened therapy (24-28 weeks)d. Consider implications before initiating PEG-IFN and RBV containing regimens.& Strong &
\\
		\end{tabularx}
		\end{center}
		\normalsize
		\vspace{10pt}
		        

\end{rSubsection}
\end{rSection}\begin{rSection}{ phenprocoumon }
\item[]
\begin{rSubsection}{ VKORC1 }{ vitamin K epoxide reductase complex, subunit 1 }{}{}
\item[]
\item[] ------------------------------------------------------ Dosing Guideline --------------------------------------------------------\newline
\item[]
\item[] \textbf{ H2/H1 } | \textbf{ H2/H1 }
\item The Royal Dutch Pharmacists Association - Pharmacogenetics Working Group has evaluated therapeutic dose recommendations for phenprocoumon based on VKORC1 genotype PMID:21412232.  They found that VKORC1 variant:rs9934438 genotype contributes to dose variability.  However, they make no dosing recommendations at this time , because of strict international normalized ratio monitoring by the Dutch Thrombosis Service., 
 \newline
\vspace{1pt}\newline
		\scriptsize
		\begin{center}
		\begin{tabularx}{0.9\textwidth}{ bssss }
		\textbf{ Genotype }&\textbf{ Therapeutic Dose Recommendation }&\textbf{ Level of Evidence }&\textbf{ Clinical Relevance }&\textbf{
}\\
		\vspace{1pt}\\
		\hline \\
		\vspace{1pt}\\
		         VKORC1 variant:rs9934438 AG & None & Published controlled studies of good quality* relating to phenotyped and/or genotyped patients or healthy volunteers, and having relevant pharmacokinetic or clinical endpoints. & Minor clinical effect (S): QTc prolongation (:450 ms , :470 ms),  INR increase : 4.5 Kinetic effect (S). &
\\
		\vspace{1pt}\\
		\hline \\
		\vspace{1pt}\\
		         VKORC1 variant:rs9934438 AA & Check INR more frequently. & Published controlled studies of good quality* relating to phenotyped and/or genotyped patients or healthy volunteers, and having relevant pharmacokinetic or clinical endpoints. & Minor clinical effect (S): QTc prolongation (:450 ms , :470 ms),  INR increase : 4.5 Kinetic effect (S). &
\\
		\end{tabularx}
		\end{center}
		\normalsize
		\vspace{10pt}
		        
\item[] ---------------------------------------------------- Clinical Annotations -----------------------------------------------------\newline
\item \textbf{\colorbox{cyan} {Class 2A}} \textbf{ rs9934438 } \textit{ GA }
\item[] Patients with the AG genotype may have decreased dose of acenocoumarol or phenprocoumon as compared to patients with genotype GG. Other genetic and clinical factors may also influence the dose of acenocoumarol or phenprocoumon.
\end{rSubsection}
\end{rSection}\begin{rSubsection}{ PRSS53 }{ protease, serine, 53 }{}{}
\item[]

\begin{rSection}{ tacrolimus }
\item[]
\begin{rSubsection}{ CYP3A5 }{ cytochrome P450, family 3, subfamily A, polypeptide 5 }{}{}
\item[]
\item[] ------------------------------------------------------ Dosing Guideline --------------------------------------------------------\newline
\item[]
\item[] \textbf{ *1A/*1A } | \textbf{ *1A/*1A }
\item July 2015
 \newline
\item Advanced online publication March 2015
 \newline
\item Patients undergoing kidney, heart, lung, or hematopoietic stem cell transplant.
 \newline
\item Patients undergoing liver transplant where the donor and recipient CYP3A5 genotypes are identical.
 \newline
\item Blood concentrations of tacrolimus are strongly influenced by CYP3A5 genotype, with substantial evidence linking CYP3A5 genotype with phenotypic variability...In kidney, heart and lung transplant patients, over 50 studies have found that individuals with the CYP3A5*1/*1 or CYP3A5*1/*3 genotype have significantly lower dose-adjusted trough concentrations of tacrolimus as compared to those with the CYP3A5*3/*3 genotype..., 
 \newline
\item Those recipients with an extensive or intermediate metabolizer phenotype will generally require an increased dose of tacrolimus to achieve therapeutic drug concentrations. We recommend a dose 1.5 - 2 times higher than standard dosing, but not to exceed 0.3 mg/kg/day, followed by therapeutic drug monitoring given the risk of arterial vasoconstriction, hypertension and nephrotoxicity that can occur with supratherapeutic tacrolimus concentrations.,  
 \newline
\item Clinical Pharmacogenetics Implementation Consortium (CPIC) Guidelines for CYP3A5 genotypes and Tacrolimus Dosing(https://github.com/PharmGKB/cpic-guidelines/raw/master/tacrolimus/2015/25801146.pdf) 
 \newline
\item 2015 Supplement(https://github.com/PharmGKB/cpic-guidelines/raw/master/tacrolimus/2015/25801146-supplement.pdf)
 \newline
\item 2015 Tacrolimus translation table(https://github.com/PharmGKB/cpic-guidelines/raw/master/tacrolimus/2015/25801146-CYP3A5\%20translation\%20table.xlsx)
 \newline
\item CYP3A5 allele frequency table(https://github.com/PharmGKB/cpic-guidelines/raw/master/tacrolimus/2015/25801146-CYP3A5\%20allele\%20frequency\%20table.xlsx)
 \newline
\item Table 1: Dosing recommendations for tacrolimus based on CYP3A5 phenotype:
 \newline
\item Adapted from Tables 1 and 2 of the 2015 guideline manuscript.
 \newline
\vspace{1pt}\newline
		\scriptsize
		\begin{center}
		\begin{tabularx}{0.9\textwidth}{ bssssss }
		\textbf{ Likely phenotype a }&\textbf{ Genotypes }&\textbf{ Examples of diplotypes b }&\textbf{ Implications for tacrolimus pharmacologic measures }&\textbf{ Therapeutic Recommendations c }&\textbf{ Classification of recommendations e }&\textbf{
}\\
		\vspace{1pt}\\
		\hline \\
		\vspace{1pt}\\
		         Extensive metabolizer (CYP3A5 expresser) & An individual carrying two functional alleles & *1/*1 & Lower dose-adjusted trough concentrations of tacrolimus and decreased chance of achieving target tacrolimus concentrations & Increase starting dose 1.5 to 2 times recommended starting dose d. Total starting dose should not exceed 0.3mg/kg/day. Use therapeutic drug monitoring to guide dose adjustments & Strong &
\\
		\vspace{1pt}\\
		\hline \\
		\vspace{1pt}\\
		         Intermediate metabolizer (CYP3A5 expresser) & An individual carrying one functional allele and one non-functional allele & *1/*3, *1/*6, *1/*7& Lower dose-adjusted trough concentrations of tacrolimus and decreased chance of achieving target tacrolimus concentrations & Increase starting dose 1.5 to 2 times recommended starting dose d. Total starting dose should not exceed 0.3mg/kg/day. Use therapeutic drug monitoring to guide dose adjustments & Strong &
\\
		\vspace{1pt}\\
		\hline \\
		\vspace{1pt}\\
		         Poor metabolizer (CYP3A5 non-expresser) & An individual carrying two non-functional alleles & *3/*3, *6/*6, *7/*7, *3/*6, *3/*7, *6/*7& Higher (“normal”) dose-adjusted trough concentrations of tacrolimus and increased chance of achieving target tacrolimus concentrations & Initiate therapy with standard recommended dose. Use therapeutic drug monitoring to guide dose adjustments & Strong &
\\
		\end{tabularx}
		\end{center}
		\normalsize
		\vspace{10pt}
		        

\end{rSubsection}

\begin{rSection}{ tegafur }
\item[]
\begin{rSubsection}{ DPYD }{ dihydropyrimidine dehydrogenase }{}{}
\item[]
\item[] ------------------------------------------------------ Dosing Guideline --------------------------------------------------------\newline
\item[]
\item[] \textbf{ *1/*1 } | \textbf{ *1/*1 }
\item May 2014 Update on PharmGKB
 \newline
\item December 2013 Publication
 \newline
\item Accepted article preview online August 2013,  Advance online publication October 2013.
 \newline
\item at the time of this writing, there are no data available on the possible role of DPYD*2A, *13, or variant:rs67376798 in 5-fluorouracil toxicities in pediatric patient populations,  however, there is no reason to suspect that DPYD variant alleles would affect 5-fluorouracil metabolism differently in children compared to adults.
 \newline
\item The strength of the dosing recommendations is based on the fact that some variants (DPYD*2A, *13, and variant:rs67376798) clearly affect DPD activity, and DPD activity is clearly related to 5-fluorouracil clearance, and 5-fluorouracil exposure is associated with its toxic effects. Therefore, reduction of fluoropyrimidine dosage in patients with these variants may prevent severe and possibly life-threatening toxicities. However, available evidence does not clearly indicate a degree of dose reduction needed to prevent fluoropyrimidine related toxicities...Based on literature review (see full manuscript), our recommendation is to start with at least a 50\% reduction of the starting dose followed by an increase in dose in patients experiencing no or clinically tolerable toxicity to maintain efficacy, a decrease in dose in patients who do not tolerate the starting dose to minimize toxicities or pharmacokinetic guided dose adjustments (if available). Patients who are homozygous for DPYD*2A, *13, or variant:rs67376798 may demonstrate complete DPD deficiency and the use of 5-fluorouracil or capecitabine is not recommended in these patients.,  
 \newline
\item Clinical Pharmacogenetics Implementation Consortium Guidelines for Dihydropyrimidine Dehydrogenase Genotype and Fluoropyrimidine Dosing(https://github.com/PharmGKB/cpic-guidelines/raw/master/fluoropyrimidines/2013/23988873.pdf)
 \newline
\item 2013 supplement(https://github.com/PharmGKB/cpic-guidelines/raw/master/fluoropyrimidines/2013/23988873-supplement.pdf)
 \newline
\item Table 1: Recommended dosing of fluoropyrimidines by genotype/phenotype.
 \newline
\item Adapted from Tables 1 and 2 of the 2013 guideline manuscript.
 \newline
\vspace{1pt}\newline
		\scriptsize
		\begin{center}
		\begin{tabularx}{0.9\textwidth}{ bsssss }
		\textbf{ Phenotype (genotype) }&\textbf{ Examples of diplotypes }&\textbf{ Implications for phenotypic measures }&\textbf{ Dosing recommendations }&\textbf{ Classification of recommendations a }&\textbf{
}\\
		\vspace{1pt}\\
		\hline \\
		\vspace{1pt}\\
		         Homozygous wild-type or normal, high DPD activity (two or more functional *1 alleles) & *1/*1 & Normal DPD activity and , normal,  risk for fluoropyrimidine toxicity & Use label-recommended dosage and administration & Moderate &
\\
		\vspace{1pt}\\
		\hline \\
		\vspace{1pt}\\
		         Heterozygous or intermediate activity (~3-5\% of patients), may have partial DPD deficiency, at risk for toxicity with drug exposure (one functional allele *1, plus one nonfunctional allele - *2A, *13 or rs67376798A c) & *1/*2A,  *1/*13,  *1/ rs67376798A c) & Decreased DPD activity (leukocyte DPD activity at 30\% to 70\% that of the normal population) and increased risk for severe or even fatal drug toxicity when treated with fluoropyrimidine drugs & Start with at least a 50\% reduction in starting dose followed by titration of dose based on toxicity b or pharmacokinetic test (if available) & Moderate &
\\
		\vspace{1pt}\\
		\hline \\
		\vspace{1pt}\\
		         Homozygous variant, DPD deficiency (~0.2\% of patients), at risk for toxicity with drug exposure (2 nonfunctional alleles - *2A, *13 or rs67376798A c) & *2A/*2A,  *13/*13,  rs67376798A c / rs67376798A c & Complete DPD deficiency and increased risk for severe or even fatal drug toxicity when treated with fluoropyrimidine drugs & Select alternate drug & Strong &
\\
		\end{tabularx}
		\end{center}
		\normalsize
		\vspace{10pt}
		        
\item[] ---------------------------------------------------- Clinical Annotations -----------------------------------------------------\newline
\item \textbf{\colorbox{red} {Class 1A}} \textbf{ rs55886062 } \textit{ AA }
\item[] Patients with the AA genotype (DPYD *1/*1) and cancer who are treated with fluoropyrimidine-based chemotherapy may have a decreased, but not absent, risk for drug toxicity as compared to patients with the AC or CC genotype (DPYD *1/*13 or *13/*13). Fluoropyrimidines are often used in combination chemotherapy such as FOLFOX (fluorouracil, leucovorin and oxaliplatin), FOLFIRI (fluorouracil, leucovorin and irinotecan) or FEC (fluorouracil, epirubicin and cyclophosphamide) or with other drugs such as bevacizumab, cetuximab, raltitrexed. The combination and delivery of the drug may influence risk for toxicity. Other genetic and clinical factors may also influence response to fluoropyrimidine-based chemotherapy.\item \textbf{\colorbox{red} {Class 1A}} \textbf{ rs3918290 } \textit{ CC }
\item[] Patients with the CC genotype (DPYD *1/*1) and cancer who are treated with fluoropyrimidine-based chemotherapy may have 1) increased clearance of fluoropyrimidine drugs and 2) decreased, but not non-existent, risk for drug toxicity as compared to patients with the CT or TT genotype (DPYD *1/*2A or *2A/*2A). Fluoropyrimidines are often used in combination chemotherapy such as FOLFOX (fluorouracil, leucovorin and oxaliplatin), FOLFIRI (fluorouracil,  leucovorin and irinotecan) or FEC (fluorouracil, epirubicin and cyclophosphamide) or with other drugs such as bevacizumab, cetuximab, raltitrexed. The combination and delivery of the drug may influence risk for toxicity. Other genetic and clinical factors may also influence response to fluoropyrimidine based chemotherapy.\item \textbf{\colorbox{red} {Class 1A}} \textbf{ rs67376798 } \textit{ TT }
\item[] Patients with the TT genotype and cancer who are treated with fluoropyrimidine-based chemotherapy may have 1) increased clearance of the drug and 2) decreased, but not absent, risk and reduced severity of drug toxicity as compared to patients with the AT genotype. Fluoropyrimidines are often used in combination chemotherapy such as FOLFOX (fluorouracil, leucovorin and oxaliplatin), FOLFIRI (fluorouracil, leucovorin and irinotecan) or FEC (fluorouracil, epirubicin and cyclophosphamide) or with other drugs such as bevacizumab, cetuximab, raltitrexed. The combination and delivery of the drug may influence risk for toxicity. Other genetic and clinical factors may also influence response to fluoropyrimidine-based chemotherapy.
\end{rSubsection}
\end{rSection}\begin{rSection}{ thalidomide }
\item[]
\begin{rSection}{ ticagrelor }
\item[]
\begin{rSection}{ tramadol }
\item[]
\begin{rSection}{ valproic acid }
\item[]
\begin{rSection}{ venlafaxine }
\item[]
\begin{rSubsection}{ CYP2D6 }{ cytochrome P450, family 2, subfamily D, polypeptide 6 }{}{}
\item[]
\item[] ------------------------------------------------------ Dosing Guideline --------------------------------------------------------\newline
\item[]
\item[] \textbf{ *1E/*1 } | \textbf{ *5/*1 }
\item The Royal Dutch Pharmacists Association - Pharmacogenetics Working Group has evaluated therapeutic dose recommendations for venlafaxine based on CYP2D6 genotypes PMID:21412232.  For PM and IM genotypes, they state that there are not sufficient data to allow calculation of dose adjustment, and they recommend selecting an alternative drug or adjusting dose to clinical response and monitoring (O-desmethyl)venlafaxine plasma concentration.  For UM genotypes, they recommend titrating dose to a maximum of 150\% of the normal dose(based on venlafaxine and (O-desmethyl)venlafaxine plasma concentration) or selecting an alternative drug.
 \newline
\vspace{1pt}\newline
		\scriptsize
		\begin{center}
		\begin{tabularx}{0.9\textwidth}{ bssss }
		\textbf{ Phenotype (Genotype) }&\textbf{ Therapeutic Dose Recommendation }&\textbf{ Level of Evidence }&\textbf{ Clinical Relevance }&\textbf{
}\\
		\vspace{1pt}\\
		\hline \\
		\vspace{1pt}\\
		         PM (two inactive (*3-*8, *11-*16, *19-*21, *38, *40, *42) alleles) & Insufficient data to allow calculation of dose adjustment.  Select alternative drug (e.g., citalopram, sertraline) or adjust dose to clinical response and monitor (O-desmethyl)venlafaxine plasma concentration. & Published controlled studies of good quality* relating to phenotyped and/or genotyped patients or healthy volunteers, and having relevant pharmacokinetic or clinical endpoints. & Clinical effect (S): long-standing discomfort (48-168 hr) without permanent injury e.g. failure of therapy with tricyclic antidepressants, atypical antipsychotic drugs,  extrapyramidal side effects,  parkinsonism,  ADE resulting from increased bioavailability of tricyclic antidepressants, metoprolol, propafenone (central effects e.g. dizziness),  INR 4.5-6.0,  neutropenia 1.0-1.5x109/l,  leucopenia 2.0-3.0x109/l,  thrombocytopenia 50-75x109/l.& 
\\
		\vspace{1pt}\\
		\hline \\
		\vspace{1pt}\\
		        IM (two decreased-activity (*9, *10, *17, *29, *36, *41) alleles or carrying one active (*1, *2, *33, *35) and one inactive (*3-*8, *11-*16, *19-*21, *38, *40, *42) allele, or carrying one decreased-activity (*9, *10, *17, *29, *36, *41) allele and one inactive (*3-*8, *11-*16, *19-*21, *38, *40, *42) allele) & Insufficient data to allow calculation of dose adjustment.  Select alternative drug (e.g., citalopram, sertraline) or adjust dose to clinical response and monitor (O-desmethyl)venlafaxine plasma concentration.& Published controlled studies of good quality* relating to phenotyped and/or genotyped patients or healthy volunteers, and having relevant pharmacokinetic or clinical endpoints. &Clinical effect (S): long-standing discomfort (48-168 hr) without permanent injury e.g. failure of therapy with tricyclic antidepressants, atypical antipsychotic drugs,  extrapyramidal side effects,  parkinsonism,  ADE resulting from increased bioavailability of tricyclic antidepressants, metoprolol, propafenone (central effects e.g. dizziness),  INR 4.5-6.0,  neutropenia 1.0-1.5x109/l,  leucopenia 2.0-3.0x109/l,  thrombocytopenia 50-75x109/l. &
\\
		\vspace{1pt}\\
		\hline \\
		\vspace{1pt}\\
		         UM (a gene duplication in absence of inactive (*3-*8, *11-*16, *19-*21, *38, *40, *42) or decreased-activity (*9, *10, *17, *29, *36, *41) alleles) & Be alert to decreased venlafaxine and increased (O-desmethyl)venlafaxine plasma concentration. Titrate dose to a maximum of 150\% of the normal dose or select alternative drug (e.g., citalopram, sertraline). & Published controlled studies of good quality* relating to phenotyped and/or genotyped patients or healthy volunteers, and having relevant pharmacokinetic or clinical endpoints. &Minor clinical effect (S): QTc prolongation (:450 ms female, :470 ms male),  INR increase : 4.5.  Kinetic effect (S).&
\\
		\end{tabularx}
		\end{center}
		\normalsize
		\vspace{10pt}
		        

\end{rSubsection}
\end{rSection}\begin{rSection}{ warfarin }
\item[]
\begin{rSubsection}{ CYP2C9 }{ cytochrome P450, family 2, subfamily C, polypeptide 9 }{}{}
\item[]
\item[] ------------------------------------------------------ Dosing Guideline --------------------------------------------------------\newline
\item[]
\item[] \textbf{ *18/*1 } | \textbf{ *3/*1 }
\item November 2013 Update
 \newline
\item CPIC guideline authors are aware of several recently published studies on warfarin pharmacogenetics PMID: 24251361, 24251363, 24251360. These papers have prompted several opinion pieces PMID: 24328463, 24251364. The authors are evaluating the information, which will be incorporated into the next update of the CPIC guideline on warfarin.
 \newline
\item October 2011
 \newline
\item Advance online publication September 2011. 
 \newline
\item adult patients
 \newline
\item Pharmacogenetic algorithm-based warfarin dosing: Numerous studies have derived warfarin dosing algorithms that use both genetic and non-genetic factors to predict warfarin dose PMID: 18305455, 19228618, 18574025. Two algorithms perform well in estimating stable warfarin dose across different ethnic populations,  PMID: 18305455, 19228618 these were created using more than 5,000 subjects. Dosing algorithms using genetics outperform nongenetic clinical algorithms and fixed-dose approaches in dose prediction PMID: 18305455, 19228618. The best way to estimate the anticipated stable dose of warfarin is to use the algorithms available on http://www.warfarindosing.org (offering both high-performing algorithms PMID: 18305455, 19228618). The dosing algorithm published by the International Warfarin Pharmacogenetics Consortium is also online, at IWPC Pharmacogenetic Dosing Algorithm(https://github.com/PharmGKB/cpic-guidelines/raw/master/warfarin/2011/IWPCdosecalculator.xls). The two algorithms provide very similar dose recommendations., 
 \newline
\item Approach to pharmacogenetic-based warfarin dosing without access to dosing algorithms:     In 2007, the FDA modified the warfarin label, stating that CYP2C9 and VKORC1 genotypes may be useful in determining the optimal initial dose of warfarin PMID:17906972. The label was further updated in 2010 to include a table (Table 1) describing recommendations for initial dosing ranges for patients with different combinations of CYP2C9 and VKORC1 genotypes. Genetics-based algorithms also better predict warfarin dose than the FDA-approved warfarin label table PMID:21272753. Therefore, the use of pharmacogenetic algorithm-based dosing is recommended when possible, although if electronic means for such dosing are not available, the table-based dosing approaches (Table 1) are suggested. The range of doses by VKORC1 genotype and the range of dose recommendations/predictions by the FDA table and algorithm are shown in Figure 2., 
 \newline
\item Clinical Pharmacogenetics Implementation Consortium Guidelines for CYP2C9 and VKORC1 genotypes and warfarin dosing(https://github.com/PharmGKB/cpic-guidelines/raw/master/warfarin/2011/21900891.pdf)
 \newline
\item 2011 supplement(https://github.com/PharmGKB/cpic-guidelines/raw/master/warfarin/2011/21900891-supplement.pdf)
 \newline
\item Look up your warfarin dosing guideline using the IWPC Pharmacogenetic Dosing Algorithm(https://github.com/PharmGKB/cpic-guidelines/raw/master/warfarin/2011/IWPCdosecalculator.xls).
 \newline
\item Figure 2. Frequency histograms of stable therapeutic warfarin doses in mg/week, stratified by VKORC1 -1639G, A genotype.
 \newline
\item Adapted from Figure 2 of the 2011 guideline manuscript
 \newline
\item !CPIC warfarin dosing guideline(Figure2-CPICWarfarindosingguidelines.png)
 \newline
\item Figure 2 Legend: Frequency histograms of stable therapeutic warfarin doses in mg/week, stratified by VKORC1 -1639G, A genotype in 3,616 patients recruited by the International Warfarin Pharmacogenetics Consortium (IWPC) who did not carry the CYP2C9*2 or *3 allele (i.e., coded as *1/*1 for US Food and Drug Administration (FDA) table and algorithm dosing). The range of doses within each genotype group recommended on the FDA table is shown via the shaded rectangle. The range of doses predicted using the IWPC dosing algorithm in these 3,616 patients is shown by the solid lines.
 \newline
\item Figure 2 demonstrates that the range of individuals covered by the FDA table is much narrower than that of the algorithm. The article and supplement detail important variables that are not covered by the table that should also be taken into consideration.
 \newline
\item Table 1: Recommended daily warfarin doses (mg/day) to achieve a therapeutic INR based on CYP2C9 and VKORC1 genotype using the warfarin product insert approved by the United States Food and Drug Administration:
 \newline
\item Adapted from Table 1 of the 2011 guideline manuscript
 \newline
\item Reproduced from updated warfarin (Coumadin) product label.
 \newline
\item Supplemental Table S1. Genotypes that constitute the * alleles for CYP2C9
 \newline
\item Adapted from Table S1 of the 2011 guideline supplement
 \newline
\vspace{1pt}\newline
		\scriptsize
		\begin{center}
		\begin{tabularx}{0.9\textwidth}{ bssss }
		\textbf{ VKORC1 Genotype (-1639G, A, variant:rs9923231) }&\textbf{ CYP2C9*1/*1 }&\textbf{ CYP2C9*1/*2 }&\textbf{ CYP2C9*1/*3 }&\textbf{ CYP2C9*2/*2 }&\textbf{ CYP2C9*2/*3 }&\textbf{ CYP2C9*3/*3}&\textbf{ 
}\\
		\vspace{1pt}\\
		\hline \\
		\vspace{1pt}\\
		         GG & 5-7 & 5-7 & 3-4 & 3-4 & 3-4 & 0.5-2&
\\
		\vspace{1pt}\\
		\hline \\
		\vspace{1pt}\\
		         GA & 5-7 & 3-4 & 3-4 & 3-4 & 0.5-2 & 0.5-2&
\\
		\vspace{1pt}\\
		\hline \\
		\vspace{1pt}\\
		         AA & 3-4  & 3-4 & 0.5-2 & 0.5-2 & 0.5-2 & 0.5-2&
\\
		\vspace{1pt}\\
		\hline \\
		\vspace{1pt}\\
		         Allele & Constituted by genotypes at:& Amino acid changes & Enzymatic Activity &
\\
		\vspace{1pt}\\
		\hline \\
		\vspace{1pt}\\
		         --- & --- & --- & --- &
\\
		\vspace{1pt}\\
		\hline \\
		\vspace{1pt}\\
		         *1 & reference allele at all positions & & Normal & 
\\
		\vspace{1pt}\\
		\hline \\
		\vspace{1pt}\\
		         *2 & C, T at variant:rs1799853 & R144C & Decreased & 
\\
		\vspace{1pt}\\
		\hline \\
		\vspace{1pt}\\
		         *3 & A, C at variant:rs1057910 & I359L & Decreased \\
		\end{tabularx}
		\end{center}
		\normalsize
		\vspace{10pt}
		        
\item[] ---------------------------------------------------- Clinical Annotations -----------------------------------------------------\newline
\item \textbf{\colorbox{red} {Class 1A}} \textbf{ rs1057910 } \textit{ AC }
\item[] Patients with the AC genotype: 1) may require a decreased dose of warfarin as compared to patients with the AA genotype 2) may have an increased risk for adverse events as compared to patients with the AA genotype.\item \textbf{\colorbox{cyan} {Class 2A}} \textbf{ rs7900194 } \textit{ GG }
\item[] Patients with the GG genotype who are treated with warfarin may require a higher maintenance dose as compared to patients with the AG or GG genotype.  Other clinical or genetic factors may also influence warfarin dose.\item \textbf{\colorbox{cyan} {Class 2A}} \textbf{ rs56165452 } \textit{ TT }
\item[] Patients with the TT genotype may required higher dose of warfarin as compared to patients with the CT or CC genotype. Other clinical or genetic factors may also influence  warfarin dose. This variant rs56165452 defines CYP2C9*4.
\end{rSubsection}
\end{rSection}\begin{rSubsection}{ VKORC1 }{ vitamin K epoxide reductase complex, subunit 1 }{}{}
\item[]
\item[] ------------------------------------------------------ Dosing Guideline --------------------------------------------------------\newline
\item[]
\item[] \textbf{ H2/H1 } | \textbf{ H2/H1 }
\item November 2013 Update
 \newline
\item CPIC guideline authors are aware of several recently published studies on warfarin pharmacogenetics PMID: 24251361, 24251363, 24251360. These papers have prompted several opinion pieces PMID: 24328463, 24251364. The authors are evaluating the information, which will be incorporated into the next update of the CPIC guideline on warfarin.
 \newline
\item October 2011
 \newline
\item Advance online publication September 2011. 
 \newline
\item adult patients
 \newline
\item Pharmacogenetic algorithm-based warfarin dosing: Numerous studies have derived warfarin dosing algorithms that use both genetic and non-genetic factors to predict warfarin dose PMID: 18305455, 19228618, 18574025. Two algorithms perform well in estimating stable warfarin dose across different ethnic populations,  PMID: 18305455, 19228618 these were created using more than 5,000 subjects. Dosing algorithms using genetics outperform nongenetic clinical algorithms and fixed-dose approaches in dose prediction PMID: 18305455, 19228618. The best way to estimate the anticipated stable dose of warfarin is to use the algorithms available on http://www.warfarindosing.org (offering both high-performing algorithms PMID: 18305455, 19228618). The dosing algorithm published by the International Warfarin Pharmacogenetics Consortium is also online, at IWPC Pharmacogenetic Dosing Algorithm(https://github.com/PharmGKB/cpic-guidelines/raw/master/warfarin/2011/IWPCdosecalculator.xls). The two algorithms provide very similar dose recommendations., 
 \newline
\item Approach to pharmacogenetic-based warfarin dosing without access to dosing algorithms:     In 2007, the FDA modified the warfarin label, stating that CYP2C9 and VKORC1 genotypes may be useful in determining the optimal initial dose of warfarin PMID:17906972. The label was further updated in 2010 to include a table (Table 1) describing recommendations for initial dosing ranges for patients with different combinations of CYP2C9 and VKORC1 genotypes. Genetics-based algorithms also better predict warfarin dose than the FDA-approved warfarin label table PMID:21272753. Therefore, the use of pharmacogenetic algorithm-based dosing is recommended when possible, although if electronic means for such dosing are not available, the table-based dosing approaches (Table 1) are suggested. The range of doses by VKORC1 genotype and the range of dose recommendations/predictions by the FDA table and algorithm are shown in Figure 2., 
 \newline
\item Clinical Pharmacogenetics Implementation Consortium Guidelines for CYP2C9 and VKORC1 genotypes and warfarin dosing(https://github.com/PharmGKB/cpic-guidelines/raw/master/warfarin/2011/21900891.pdf)
 \newline
\item 2011 supplement(https://github.com/PharmGKB/cpic-guidelines/raw/master/warfarin/2011/21900891-supplement.pdf)
 \newline
\item Look up your warfarin dosing guideline using the IWPC Pharmacogenetic Dosing Algorithm(https://github.com/PharmGKB/cpic-guidelines/raw/master/warfarin/2011/IWPCdosecalculator.xls).
 \newline
\item Figure 2. Frequency histograms of stable therapeutic warfarin doses in mg/week, stratified by VKORC1 -1639G, A genotype.
 \newline
\item Adapted from Figure 2 of the 2011 guideline manuscript
 \newline
\item !CPIC warfarin dosing guideline(Figure2-CPICWarfarindosingguidelines.png)
 \newline
\item Figure 2 Legend: Frequency histograms of stable therapeutic warfarin doses in mg/week, stratified by VKORC1 -1639G, A genotype in 3,616 patients recruited by the International Warfarin Pharmacogenetics Consortium (IWPC) who did not carry the CYP2C9*2 or *3 allele (i.e., coded as *1/*1 for US Food and Drug Administration (FDA) table and algorithm dosing). The range of doses within each genotype group recommended on the FDA table is shown via the shaded rectangle. The range of doses predicted using the IWPC dosing algorithm in these 3,616 patients is shown by the solid lines.
 \newline
\item Figure 2 demonstrates that the range of individuals covered by the FDA table is much narrower than that of the algorithm. The article and supplement detail important variables that are not covered by the table that should also be taken into consideration.
 \newline
\item Table 1: Recommended daily warfarin doses (mg/day) to achieve a therapeutic INR based on CYP2C9 and VKORC1 genotype using the warfarin product insert approved by the United States Food and Drug Administration:
 \newline
\item Adapted from Table 1 of the 2011 guideline manuscript
 \newline
\item Reproduced from updated warfarin (Coumadin) product label.
 \newline
\item Supplemental Table S1. Genotypes that constitute the * alleles for CYP2C9
 \newline
\item Adapted from Table S1 of the 2011 guideline supplement
 \newline
\vspace{1pt}\newline
		\scriptsize
		\begin{center}
		\begin{tabularx}{0.9\textwidth}{ bssss }
		\textbf{ VKORC1 Genotype (-1639G, A, variant:rs9923231) }&\textbf{ CYP2C9*1/*1 }&\textbf{ CYP2C9*1/*2 }&\textbf{ CYP2C9*1/*3 }&\textbf{ CYP2C9*2/*2 }&\textbf{ CYP2C9*2/*3 }&\textbf{ CYP2C9*3/*3}&\textbf{ 
}\\
		\vspace{1pt}\\
		\hline \\
		\vspace{1pt}\\
		         GG & 5-7 & 5-7 & 3-4 & 3-4 & 3-4 & 0.5-2&
\\
		\vspace{1pt}\\
		\hline \\
		\vspace{1pt}\\
		         GA & 5-7 & 3-4 & 3-4 & 3-4 & 0.5-2 & 0.5-2&
\\
		\vspace{1pt}\\
		\hline \\
		\vspace{1pt}\\
		         AA & 3-4  & 3-4 & 0.5-2 & 0.5-2 & 0.5-2 & 0.5-2&
\\
		\vspace{1pt}\\
		\hline \\
		\vspace{1pt}\\
		         Allele & Constituted by genotypes at:& Amino acid changes & Enzymatic Activity &
\\
		\vspace{1pt}\\
		\hline \\
		\vspace{1pt}\\
		         --- & --- & --- & --- &
\\
		\vspace{1pt}\\
		\hline \\
		\vspace{1pt}\\
		         *1 & reference allele at all positions & & Normal & 
\\
		\vspace{1pt}\\
		\hline \\
		\vspace{1pt}\\
		         *2 & C, T at variant:rs1799853 & R144C & Decreased & 
\\
		\vspace{1pt}\\
		\hline \\
		\vspace{1pt}\\
		         *3 & A, C at variant:rs1057910 & I359L & Decreased \\
		\end{tabularx}
		\end{center}
		\normalsize
		\vspace{10pt}
		        
\item[] ---------------------------------------------------- Clinical Annotations -----------------------------------------------------\newline
\item \textbf{\colorbox{orange} {Class 1B}} \textbf{ rs9934438 } \textit{ GA }
\item[] Patients with the AG genotype who are treated with warfarin may require a lower dose as compared to patients with the GG genotype, and a higher dose as compared to patients with the AA genotype. Other clinical and genetic factors may also influence a patient’s required dose of warfarin. \item \textbf{\colorbox{cyan} {Class 2A}} \textbf{ rs9923231 } \textit{ CT }
\item[] Patients with genotype CT may require shorter time to therapeutic INR when treated with warfarin as compared with patients with genotype CC. Other genetic and clinical factors may also influence the response to warfarin. \item \textbf{\colorbox{cyan} {Class 2A}} \textbf{ rs9923231 } \textit{ CT }
\item[] Patients with the CT genotype may have increased risk of over-anticoagulation when treated with warfarin as compared with patients with genotype CC. Other genetic and clinical factors may also influence the toxicity to warfarin.\item \textbf{\colorbox{blue} {Class 2B}} \textbf{ rs7196161 } \textit{ GA }
\item[] Patients with the AG genotype may require an increased dose of warfarin as compared to patients with the GG genotype and a decreased dose of warfarin as compared to patients with the AA genotype. Other clinical and genetic factors may also influence the dose of warfarin. 
\end{rSection}
\end{document}